### PR TITLE
Move text from tactics.rst to ltac.rst

### PR DESCRIPTION
**Pull request by jfehrle**
Attn: @Zimmi48

### CODE REVIEW

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst:1671]: The move of the section is incomplete, the section continues from `is_evar` until `is_var`. In #15015, I've noticed that you've also appended some related tactics after that (`is_cofix`, etc.). These should also be added here, rather than there.

**Comment by jfehrle** [comment]
Updated.

**Review by Zimmi48 — COMMENTED** [review]
I think all the `.. exn::` can be safely indented because they are all related to the tactic described just above.

We could take advantage of this PR to document the newly introduced tactics (`is_cofix` to `is_proj`) or do that later. cc @JasonGross @MSoegtropIMC in case you want to contribute a description for these.

**Comment by jfehrle** [comment]
> I think all the `.. exn::` can be safely indented because they are all related to the tactic described just above.

Done, updated.

**Review by Zimmi48 — APPROVED** [review]
I will merge as is if there are no suggestions to document the newly added tactics.

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/ltac.rst:1738]:
```suggestion
   :name: is_cofix
   
   Checks whether its argument is a bare cofixpoint.
   
   .. exn:: not a cofix definition
      :undocumented:
```

Here's an example, if you want to work it in:
```coq
Require Import Coq.Lists.Streams.
Goal True.
  let c := constr:(cofix f : Stream unit := Cons tt f) in
  is_cofix c.
```

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/ltac.rst:1747]:
```suggestion
   :name: is_fix
   
   Checks whether its argument is a bare fixpoint.
   
   .. exn:: not a fix definition.
      :undocumented:
```

Here's an example, if you want to work it in:
```coq
Goal True.
  is_fix (fix f (n : nat) := match n with S n => f n | O => O end).
```

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/ltac.rst:1741]:
```suggestion
   :name: is_const
   
   Checks whether its argument is a global constant.
   
   .. exn:: not a constant.
      :undocumented:
```

Here's an example, if you want to work it in:
```coq
Goal True.
  is_const Nat.add.
```

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/ltac.rst:1744]:
```suggestion
   :name: is_constructor
   
   Checks whether its argument is the constructor of a (co)inductive type.
   
   .. exn:: not a constructor.
      :undocumented:
```

Here's an example, if you want to work it in:
```coq
Goal True.
  is_constructor @pair.
```

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/ltac.rst:1753]:
```suggestion
   :name: is_ind
   
   Checks whether its argument is a (co)inductive type (family).
   
   .. exn:: not an (co)inductive datatype.
      :undocumented:
```

Here's an example, if you want to work it in:
```coq
Goal True.
  is_ind prod.
```
Note that `is_ind (list nat)`, for example, fails.

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/ltac.rst:1750]: As far as I can tell, this is equivalent to `assert_fails (has_evar @one_term)`, but I'm not completely sure.

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/ltac.rst:1755]:
```suggestion
   :name: is_proj
   
   Checks whether its argument is a primitive projection applied to a record argument.
   
   .. exn:: not a primitive projection.
      :undocumented:
```

Here's an example, if you want to work it in:
```coq
Set Primitive Projections.
Record Box {T : Type} := box { unbox : T }.
Arguments box {_} _.
Goal True.
  is_proj (unbox (box 0)).
```

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/ltac.rst:1756]:
```suggestion
   :name: is_proj
   
   Checks whether its argument is a primitive projection applied to a record argument.
   
   .. exn:: not a primitive projection.
      :undocumented:
```

Here's an example, if you want to work it in:
```coq
Set Primitive Projections.
Record Box {T : Type} := box { unbox : T }.
Arguments box {_} _.
Goal True.
  is_proj (unbox (box 0)).
```

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/ltac.rst:1750]: And the error message is "Not ground.", and can be obtained by, e.g., `is_ground open_constr:(?[c]).`, while something like `is_ground 0` succeeds.

File: doc/sphinx/proof-engine/ltac.rst
Comment (jfehrle) [reply doc/sphinx/proof-engine/ltac.rst:1741]: Note that `.. tacn:: is_const @one_term` sets the name to `is_const`, so `:name:` isn't needed here.  This works for one or more alphabetic words identifying a tactic, command, flag, etc.

File: doc/sphinx/proof-engine/ltac.rst
Comment (jfehrle) [reply doc/sphinx/proof-engine/ltac.rst:1747]: "bare" meaning that the top node of the AST is a fix construct?

**Comment by jfehrle** [comment]
Updated.

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/ltac.rst]: Yes

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/ltac.rst:1738]: Nit:
```suggestion
   Succeeds if :n:`@one_term` is a global constant that is neither a (co)inductive
```

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/ltac.rst:1730]:
```suggestion
   Succeeds if :n:`@term` is a variable or hypothesis in
```
Or is `is_var` also taking `@one_term` despite what it says here?

File: doc/sphinx/proof-engine/ltac.rst
Comment (jfehrle) [reply doc/sphinx/proof-engine/ltac.rst]: Can you confirm that a `let fix` will fail?  I'm adding a hyperlink to this nonterminal and if it does fail, I want to mention that `let fix` fails.

image

File: doc/sphinx/proof-engine/ltac.rst
Comment (jfehrle) [reply doc/sphinx/proof-engine/ltac.rst:1784]: This is the underlying code.  Is there a user-visible difference between defined and undefined evars?  The "undefined" distinction may just be internals.  Perhaps you could construct an example to test whether there's a difference?  If they are different, I'd like to explain that very briefly.

```
let has_undefined_evars evd t =
  let rec has_ev t =
    match EConstr.kind evd t with
    | Evar _ -> raise NotInstantiatedEvar
    | _ -> EConstr.iter evd has_ev t in
  try let _ = has_ev t in false
  with (Not_found | NotInstantiatedEvar) -> true

let is_ground_term evd t =
  not (has_undefined_evars evd t) 
```

Comment (JasonGross) [reply doc/sphinx/proof-engine/ltac.rst:1784]: `is_ground` is in fact implemented as the negation of `has_evar`.  The "undefined" part is just in naming, and is implicit in the fact that we're using EConstr, where all observable evars are undefined.  (In Constr-land we can have evars that happen to be defined already.)
https://github.com/coq/coq/blob/cf60242ed3022c332d3de88fbcbab4587cda83b5/plugins/ltac/internals.ml#L299-L303
https://github.com/coq/coq/blob/cf60242ed3022c332d3de88fbcbab4587cda83b5/tactics/class_tactics.ml#L1389-L1393
https://github.com/coq/coq/blob/cf60242ed3022c332d3de88fbcbab4587cda83b5/engine/evarutil.ml#L99-L108

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/ltac.rst]: `let (co)fix name body in term` is basically just syntactic sugar for `let name := (co)fix name body in term`, in the same way that `if` and destructuring `let` are just syntactic sugar for `match`.  Examples:
```coq
Goal True.
  is_fix (let fix f (n : nat) := match n with S n => f n | O => O end in f).
  (* Error: Tactic failure: not a fix definition. *)
```
```coq
Require Import Coq.Lists.Streams.
Goal True.
  is_cofix (let cofix f : Stream unit := Cons tt f in f).
  (* Error: Tactic failure: not a cofix definition. *)
```

File: doc/sphinx/proof-engine/ltac.rst
Comment (jfehrle) [reply doc/sphinx/proof-engine/ltac.rst]: `is_ind list` succeeds, so it's odd that it fails for `list nat`, which I would think is an inductive type.  Is this correct behavior for `list nat`?  Is there an understandable explanation (though I could just note that it doesn't work)?

**Comment by jfehrle** [comment]
Updated, waiting for one more detail on `is_ind`.

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/ltac.rst]: Yes, the behavior is correct.  All of these tactics are about the top-most / root node of the AST.  The root node of `list nat` is an application, not an inductive.

**Comment by jfehrle** [comment]
OK, if there are not further comments, please merge.

**Review by JasonGross — APPROVED** [review]

**Comment by JasonGross** [comment]
Feel free to take my suggestions or leave them; I'm happy to have this merged as is, if desired.

File: doc/sphinx/proof-engine/ltac.rst
Comment (jfehrle) [reply doc/sphinx/proof-engine/ltac.rst:1673]: It is the sound of one hand clapping :-)

File: doc/sphinx/proof-engine/ltac.rst
Comment (jfehrle) [reply doc/sphinx/proof-engine/ltac.rst:1706]: This looks like `core` is the answer, but I don't understand how the hint database or transparency info is relevant.  Maybe a bad explanation?

```
let unify ?(state=TransparentState.full) x y =
  Proofview.Goal.enter begin fun gl ->
  let env = Proofview.Goal.env gl in
  let sigma = Proofview.Goal.sigma gl in
  try
    let core_flags =
      { (default_unify_flags ()).core_unify_flags with
        modulo_delta = state;
        modulo_conv_on_closed_terms = Some state} in
    (* What to do on merge and subterm flags?? *)
    let flags = { (default_unify_flags ()) with
      core_unify_flags = core_flags;
      merge_unify_flags = core_flags;
      subterm_unify_flags = { core_flags with modulo_delta = TransparentState.empty } }
    in
    let sigma = w_unify (Tacmach.pf_env gl) sigma Reduction.CONV ~flags x y in
    Proofview.Unsafe.tclEVARS sigma
  with e when noncritical e ->
    let e, info = Exninfo.capture e in
    Proofview.tclZERO ~info (PretypeError (env, sigma, CannotUnify (x, y, None)))
  end
```

**Comment by Zimmi48** [comment]
@jfehrle I'm waiting on your confirmation to go ahead and merge this.

File: doc/sphinx/proof-engine/ltac.rst
Comment (SkySkimmer) [reply doc/sphinx/proof-engine/ltac.rst:1673]: This one checks if the arguments are equal up to adding some universe equations.
_strict checks if they are equal with the current universe equations.
_nounivs just ignores universes.

So I would say something like
~~~
constr_eq_strict

Succeeds if the arguments are equal modulo alpha conversion and ignoring casts. Universes are considered equal when they are equal in the universe graph.

constr_eq

Like constr_eq_strict, but may add constraints to make universes equal.

constr_eq_nounivs

Like constr_eq_strict, but all universes are equated.
~~~

Comment (SkySkimmer) [reply doc/sphinx/proof-engine/ltac.rst:1673]: also IIRC _nounivs is the one used by ltac `match`, eg 
~~~coq
  let c := constr:(Type) in
  let x := constr:(Type : c) in
  match c with
  | x => idtac c x
  end.
~~~
succeeds.

File: doc/sphinx/proof-engine/ltac.rst
Comment (SkySkimmer) [reply doc/sphinx/proof-engine/ltac.rst:1706]: The `core` in that code is unrelated to the database name. The default transparency is from `?(state=TransparentState.full)` ie everything is transparent.

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/ltac.rst:1673]: @SkySkimmer I like your suggested text, though I'd suggest "but all universes are considered equal." instead of "but all universes are equated." for `constr_eq_nounivs`

File: doc/sphinx/proof-engine/ltac.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/ltac.rst:1706]: @jfehrle Here is an example showing how hint databases can be used to adjust unification:
```coq
Create HintDb foo.
Definition idnat1 (n : nat) := n.
Definition idnat2 (n : nat) := n.
Goal True.
  unify idnat1 idnat2 with foo. (* success, by unfolding idnat1 and idnat2 *)
  Hint Opaque idnat1 : foo.
  (* Error: Ltac call to "unify (constr) (constr) with (preident)" failed.
       Unable to unify "idnat1" with "idnat2". *)
```

Comment (jfehrle) [reply doc/sphinx/proof-engine/ltac.rst]: OK, I said this:

image

**Comment by jfehrle** [comment]
Updated.  If no further comments within 24 hours, please merge.

**Comment by Zimmi48** [comment]
@coqbot merge now

**Pull request merged**
